doc/man-source/irw.1
.TH IRW "1" "Last change: Sep 2015" "irw @version@" "User Commands"
.SH NAME
.P
\fBirw\fR - Send data from Unix domain socket to stdout
.SH SYNOPSIS
.P
\fBirw\fR [\fIsocket\fR]
.SH DESCRIPTION

\fBirw\fR will connect to any Unix domain socket and print the
received data to stdout. Without a \fIsocket\fR argument it will watch
\fI@varrundir@/lirc/lircd\fR. Useful for debugging.
.P
Options:
.TP
\fB\-h\fR \fI\-\-help\fR
Display usage summary
.TP
\fB\-v\fR \fI\-\-version\fR
Display version
.SH SIGNALS
.P
On receiving SIGUSR1 \fBirw\fR makes a clean exit.

.SH EXAMPLE
.P
Dumping the default socket while pressing one button:

.nf

    myhost $ irw
    0000000000f40bf0 00 KEY_POWER ANIMAX
    0000000000f40bf0 01 KEY_POWER ANIMAX
    0000000000f40bf0 02 KEY_POWER ANIMAX
    0000000000f40bf0 03 KEY_POWER ANIMAX
    0000000000f40bf0 04 KEY_POWER ANIMAX
    0000000000f40bf0 05 KEY_POWER ANIMAX
    ^C
    myhost $

.if

.P

Note that the second counter is incremented for each new event generated
by the single key press. This is important for correct lircd operation.

.SH NOTES
.P
Since the lircd socket supports multiple clients, \fBirw\fR can run in
parallel with other applications.
.P
\fBirw\fR does nothing more than dumps a unix-domain socket, a task which
many standard tools also can do. E. g., the ncat(1) tool can be
used instead of irw to provide more flexibility (but without a sane default
socket):
.nf
    ncat -U @varrundir@/lirc/lircd
.fi

.SH "SEE ALSO"
.P
The documentation for \fBlirc\fR
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.4-devel
  * The caraca driver has been retired (still available in attic/).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * Enhanced lircd-setup tool (#123)
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers.
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * devinput.lircd.conf is generated at build time to match host kernel.
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).

  * CLI options for mode2, xmode2 and irrecord has been streamlined.
    All tools falls back to lircd defaults where it makes sense.
  * New logging API with hooks for enhanced messages and filtering
    (also #129).
  * The out-of-tree build interface is enhanced with hooks for external
    drivers to rebuild documentation and add a config file.
  * irrecord has benn taught to handle remotes without gap (#98).
  * Add an "auto" device to devinput driver for single-device
    usecase (#134).

  * Fix buggy handling of not decoded buttons in irrecord (#146).
  * Fix --plugindir not being respected (#144).
  * Fix crash for non-existing device (#145).
  * Fix gid and supplementary groups in CLI apps dropping root.
  * Fix a crash in irrecord when invoked without options.
  * Don't require write access if not needed in default.c (#133).
  * Clean up docs, remove circular links (#136). Documentation is
    now only in lirc.org/html.

0.9.3  4/09/15
  * Digest of 0.9.3pre1..0.9.3pre3

  * New build dependency: xsltproc (sometimes coming with libxml).
  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
     + Minor fixes to irrecord-ui including -k option parsing fixed.
  * mode2 has been updated
     +  handle also LIRCCODE devices (#94),
     +  Drop root privileges (#91)
     +  Add  new scope-like display
     +  Handle the--driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Create log directory if required (#124).
  * lirc-setup: Fixed name clash on recent python versions.

  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available "out there".

  * The library routines for parsing lircd.conf now warns for
    duplicate keys and values (#107).
  * Enhanced pkg-config support for drivers compiled out-of-tree
    (see driver-api in the manual).
  * Hardcoded CFLAGS in Makefiles fixed (#124).
  * The compiler has been switched to c++, and filenames are
    updated to *.cpp (#79)
  * Build: now uses parallel make by default.
  * Overall code formatting is improved with more consistent
    layout and updated git hook. See CONTRIBUTE.md

  * Unlinked and hard to find driver docs and manpages have
    been linked.
  * The documentation is heavily revised. Old obsolete text is
    removed, the manual is reorganized, some info is moved to
    manpages and a lot of minor updates and bugfixes.

  * The lircd.simsend and lircd.simreceive tools have been removed,
    superseeded by irsimsend and irsimreceive.
  * A new kernel driver irpipe has been added. This allows feeding data
    from userspace to lircd, which can use the /dev/irpipeX device
    as a clean substitute for /dev/lircX. A companion CLI tool
    irpipe(1) is also added.
  * Options parsing (notably -U) and just listing first remote
    bugs in irsimsend have been fixed.


0.9.3pre3 16/08/15
  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * Create log directory if required (#124).
  * The remotes/ directory is not used any more, removed.
  * The lircd.simsend and lircd.simreceive tools have been removed,
    superseeded by irsimsend and irsimreceive.
  * Minor fixes to lirc-setup (bbd4ccf, e384375).
  * Minor fixes to irrecord-ui including -k option parsing fixed.
  * Hardcoded CFLAGS in Makefiles fixed (#124).
  * Enhanced pkg-config support for drivers compiled out-of-tree
    (see driver-api in the manual).


tools/irw.c
/****************************************************************************
** irw.c *******************************************************************
****************************************************************************
*
* irw - watch the codes as lircd recognize them
*
* Copyright (C) 1998 Trent Piepho <xyzzy@u.washington.edu>
* Copyright (C) 1998 Christoph Bartelmus <lirc@bartelmus.de>
*
*/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <errno.h>
#include <getopt.h>

#include "lirc_log.h"

static struct option long_options[] = {
	{ "help",    no_argument, NULL, 'h' },
	{ "version", no_argument, NULL, 'v' },
	{ 0,	     0,		  0,	0   }
};

int main(int argc, char* argv[])
{
	int fd, i;
	char buf[128];
	struct sockaddr_un addr;
	int c;
	char* progname;

	progname = "irw " VERSION;

	addr.sun_family = AF_UNIX;

	while ((c = getopt_long(argc, argv, "hv", long_options, NULL))
	       != EOF) {
		switch (c) {
		case 'h':
			printf("Usage: %s [socket]\n", argv[0]);
			printf("\t -h --help \t\tdisplay usage summary\n");
			printf("\t -v --version \t\tdisplay version\n");
			return EXIT_SUCCESS;
		case 'v':
			printf("%s\n", progname);
			return EXIT_SUCCESS;
		case '?':
			fprintf(stderr, "unrecognized option: -%c\n", optopt);
			fprintf(stderr, "Try `%s --help' for more information.\n", progname);
			return EXIT_FAILURE;
		}
	}
	if (argc == optind) {
		/* no arguments */
		strcpy(addr.sun_path, LIRCD);
	} else if (argc == optind + 1) {
		/* one argument */
		strncpy(addr.sun_path, argv[optind], sizeof(addr.sun_path) - 1);
	} else {
		fprintf(stderr, "%s: incorrect number of arguments.\n", progname);
		fprintf(stderr, "Try `%s --help' for more information.\n", progname);
		return EXIT_FAILURE;
	}

	fd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (fd == -1) {
		perror("socket");
		exit(errno);
	}
	;
	if (connect(fd, (struct sockaddr*)&addr, sizeof(addr)) == -1) {
		perror("connect");
		exit(errno);
	}
	;

	for (;; ) {
		i = read(fd, buf, 128);
		if (i == -1) {
			perror("read");
			exit(errno);
		}
		;
		if (!i)
			exit(0);
		chk_write(STDOUT_FILENO, buf, i);
	}
	;
}
